### PR TITLE
Deny broken intra-doc links in linkchecker

### PR DESCRIPTION
**Pull request by jyn514**
Since rustdoc isn't warning about these links, check for them manually.

This also fixes the broken links that popped up from the lint.

### CODE REVIEW

**Comment by jyn514** [comment]
This caught quite a lot of broken links:

<details>

```
std/option/enum.Option.html:216: broken intra-doc link - [<code>Some(t)</code>]
std/iter/trait.DoubleEndedIterator.html:87: broken intra-doc link - [<code>Ok(())</code>]
std/iter/trait.DoubleEndedIterator.html:88: broken intra-doc link - [<code>Err(k)</code>]
std/iter/trait.DoubleEndedIterator.html:91: broken intra-doc link - [<code>Ok(())</code>]
std/ffi/struct.CStr.html:205: broken intra-doc link - [<code>&amp;str</code>]
std/ffi/struct.CStr.html:206: broken intra-doc link - [<code>&amp;str</code>]
std/ffi/struct.CString.html:355: broken intra-doc link - [<code>&amp;str</code>]
std/ffi/struct.CString.html:356: broken intra-doc link - [<code>&amp;str</code>]
std/ops/struct.RangeInclusive.html:253: broken intra-doc link - [<code>begin</code>, <code>end</code>]
std/slice/trait.SliceIndex.html:76: broken intra-doc link - [<code>begin</code>, <code>end</code>]
std/alloc/trait.AllocRef.html:68: broken intra-doc link - [<code>NonNull&lt;[u8]&gt;</code>]
std/alloc/trait.AllocRef.html:71: broken intra-doc link - [<code>NonNull&lt;[u8]&gt;</code>]
std/alloc/trait.AllocRef.html:96: broken intra-doc link - [<code>NonNull&lt;[u8]&gt;</code>]
std/alloc/trait.AllocRef.html:105: broken intra-doc link - [<code>NonNull&lt;[u8]&gt;</code>]
std/alloc/trait.AllocRef.html:148: broken intra-doc link - [<code>NonNull&lt;[u8]&gt;</code>]
std/alloc/trait.AllocRef.html:157: broken intra-doc link - [<code>NonNull&lt;[u8]&gt;</code>]
std/convert/trait.AsRef.html:30: broken intra-doc link - [<code>&amp;str</code>]
std/convert/trait.AsRef.html:31: broken intra-doc link - [<code>&amp;str</code>]
unstable-book/print.html:4209: broken intra-doc link - [<code>Default::default()</code>]
unstable-book/library-features/default-free-fn.html:170: broken intra-doc link - [<code>Default::default()</code>]
reference/print.html:298: broken intra-doc link - [<code>b</code> <code>B</code>]
reference/print.html:299: broken intra-doc link - [<code>a</code>-<code>z</code>]
reference/print.html:300: broken intra-doc link - [<code>b</code> <code>B</code>]
reference/print.html:440: broken intra-doc link - [<code>a</code>-<code>z</code> <code>A</code>-<code>Z</code>] [<code>a</code>-<code>z</code> <code>A</code>-<code>Z</code> <code>0</code>-<code>9</code> <code>_</code>]
reference/print.html:441: broken intra-doc link - [<code>a</code>-<code>z</code> <code>A</code>-<code>Z</code> <code>0</code>-<code>9</code> <code>_</code>]
reference/print.html:467: broken intra-doc link - [<code>/</code> <code>!</code>]
reference/print.html:470: broken intra-doc link - [<code>*</code> <code>!</code>]
reference/print.html:885: broken intra-doc link - [<code>0</code>-<code>1</code>]
reference/print.html:886: broken intra-doc link - [<code>0</code>-<code>7</code>]
reference/print.html:887: broken intra-doc link - [<code>0</code>-<code>9</code>]
reference/print.html:888: broken intra-doc link - [<code>0</code>-<code>9</code> <code>a</code>-<code>f</code> <code>A</code>-<code>F</code>]
reference/comments.html:172: broken intra-doc link - [<code>/</code> <code>!</code>]
reference/comments.html:175: broken intra-doc link - [<code>*</code> <code>!</code>]
reference/identifiers.html:172: broken intra-doc link - [<code>a</code>-<code>z</code> <code>A</code>-<code>Z</code>] [<code>a</code>-<code>z</code> <code>A</code>-<code>Z</code> <code>0</code>-<code>9</code> <code>_</code>]
reference/identifiers.html:173: broken intra-doc link - [<code>a</code>-<code>z</code> <code>A</code>-<code>Z</code> <code>0</code>-<code>9</code> <code>_</code>]
reference/tokens.html:459: broken intra-doc link - [<code>0</code>-<code>1</code>]
reference/tokens.html:460: broken intra-doc link - [<code>0</code>-<code>7</code>]
reference/tokens.html:461: broken intra-doc link - [<code>0</code>-<code>9</code>]
reference/tokens.html:462: broken intra-doc link - [<code>0</code>-<code>9</code> <code>a</code>-<code>f</code> <code>A</code>-<code>F</code>]
reference/notation.html:181: broken intra-doc link - [<code>b</code> <code>B</code>]
reference/notation.html:182: broken intra-doc link - [<code>a</code>-<code>z</code>]
reference/notation.html:183: broken intra-doc link - [<code>b</code> <code>B</code>]
embedded-book/print.html:4208: broken intra-doc link - [<code>std::process::Command</code>]
embedded-book/interoperability/c-with-rust.html:233: broken intra-doc link - [<code>std::process::Command</code>]
alloc/slice/trait.SliceIndex.html:34: broken intra-doc link - [<code>begin</code>, <code>end</code>]
alloc/alloc/trait.AllocRef.html:68: broken intra-doc link - [<code>NonNull&lt;[u8]&gt;</code>]
alloc/alloc/trait.AllocRef.html:71: broken intra-doc link - [<code>NonNull&lt;[u8]&gt;</code>]
alloc/alloc/trait.AllocRef.html:96: broken intra-doc link - [<code>NonNull&lt;[u8]&gt;</code>]
alloc/alloc/trait.AllocRef.html:105: broken intra-doc link - [<code>NonNull&lt;[u8]&gt;</code>]
alloc/alloc/trait.AllocRef.html:148: broken intra-doc link - [<code>NonNull&lt;[u8]&gt;</code>]
alloc/alloc/trait.AllocRef.html:157: broken intra-doc link - [<code>NonNull&lt;[u8]&gt;</code>]
error-index.html:10959: broken intra-doc link - [<code>llvm_asm</code>]
error-index.html:10969: broken intra-doc link - [<code>llvm_asm</code>]
error-index.html:10982: broken intra-doc link - [<code>llvm_asm</code>]
error-index.html:10995: broken intra-doc link - [<code>llvm_asm</code>]
error-index.html:11008: broken intra-doc link - [<code>llvm_asm</code>]
core/option/enum.Option.html:216: broken intra-doc link - [<code>Some(t)</code>]
core/iter/trait.DoubleEndedIterator.html:87: broken intra-doc link - [<code>Ok(())</code>]
core/iter/trait.DoubleEndedIterator.html:88: broken intra-doc link - [<code>Err(k)</code>]
core/iter/trait.DoubleEndedIterator.html:91: broken intra-doc link - [<code>Ok(())</code>]
core/ops/struct.RangeInclusive.html:250: broken intra-doc link - [<code>begin</code>, <code>end</code>]
core/slice/trait.SliceIndex.html:76: broken intra-doc link - [<code>begin</code>, <code>end</code>]
core/alloc/trait.AllocRef.html:68: broken intra-doc link - [<code>NonNull&lt;[u8]&gt;</code>]
core/alloc/trait.AllocRef.html:71: broken intra-doc link - [<code>NonNull&lt;[u8]&gt;</code>]
core/alloc/trait.AllocRef.html:96: broken intra-doc link - [<code>NonNull&lt;[u8]&gt;</code>]
core/alloc/trait.AllocRef.html:105: broken intra-doc link - [<code>NonNull&lt;[u8]&gt;</code>]
core/alloc/trait.AllocRef.html:148: broken intra-doc link - [<code>NonNull&lt;[u8]&gt;</code>]
core/alloc/trait.AllocRef.html:157: broken intra-doc link - [<code>NonNull&lt;[u8]&gt;</code>]
core/convert/trait.AsRef.html:30: broken intra-doc link - [<code>&amp;str</code>]
core/convert/trait.AsRef.html:31: broken intra-doc link - [<code>&amp;str</code>]
cargo/print.html:3528: broken intra-doc link - [<code>cargo rustdoc</code>]
cargo/reference/environment-variables.html:194: broken intra-doc link - [<code>cargo rustdoc</code>]
thread 'main' panicked at 'found some broken links', src/tools/linkchecker/main.rs:73:9
```

</details>

**Comment by jyn514** [comment]
The ones in the reference seem to be intentional, they're ~EBNF and marking literals as code blocks. I'll ignore them in the linkchecker.

**Comment by jyn514** [comment]
This is now bumped on a cargo update, https://github.com/rust-embedded/book/pull/271 merged, and an `embedded-book` update. Otherwise it's ready for review though.

File: compiler/rustc_error_codes/src/error_codes/E0660.md [hunk 1]
```diff
@@ -9,4 +9,4 @@ llvm_asm!("nop" "nop");
 Considering that this would be a long explanation, we instead recommend you
 take a look at the [`llvm_asm`] chapter of the Unstable book:
 
-[llvm_asm]: https://doc.rust-lang.org/stable/unstable-book/library-features/llvm-asm.html
+[`llvm_asm`]: https://doc.rust-lang.org/stable/unstable-book/library-features/llvm-asm.html
```
Comment (GuillaumeGomez) [inline compiler/rustc_error_codes/src/error_codes/E0660.md:12]: Why this change?

File: compiler/rustc_error_codes/src/error_codes/E0661.md [hunk 2]
```diff
@@ -10,4 +10,4 @@ llvm_asm!("nop" : "r"(a));
 Considering that this would be a long explanation, we instead recommend you
 take a look at the [`llvm_asm`] chapter of the Unstable book:
 
-[llvm_asm]: https://doc.rust-lang.org/stable/unstable-book/library-features/llvm-asm.html
+[`llvm_asm`]: https://doc.rust-lang.org/stable/unstable-book/library-features/llvm-asm.html
```
Comment (GuillaumeGomez) [inline compiler/rustc_error_codes/src/error_codes/E0661.md:13]: Why this change?

File: compiler/rustc_error_codes/src/error_codes/E0662.md [hunk 3]
```diff
@@ -13,4 +13,4 @@ llvm_asm!("xor %eax, %eax"
 Considering that this would be a long explanation, we instead recommend you
 take a look at the [`llvm_asm`] chapter of the Unstable book:
 
-[llvm_asm]: https://doc.rust-lang.org/stable/unstable-book/library-features/llvm-asm.html
+[`llvm_asm`]: https://doc.rust-lang.org/stable/unstable-book/library-features/llvm-asm.html
```
Comment (GuillaumeGomez) [inline compiler/rustc_error_codes/src/error_codes/E0662.md:16]: Bis repetita. ;)

File: compiler/rustc_error_codes/src/error_codes/E0660.md (hunk 1 above)
Comment (jyn514) [reply compiler/rustc_error_codes/src/error_codes/E0660.md:12]: The link was broken before. The reference text has to match _exactly_ or it won't be treated as a link.

Comment (jyn514) [reply compiler/rustc_error_codes/src/error_codes/E0660.md:12]: https://github.com/rust-lang/rust/issues/62345

File: src/tools/linkchecker/main.rs [hunk 4]
```diff
@@ -50,6 +53,44 @@ const LINKCHECK_EXCEPTIONS: &[(&str, &[&str])] = &[
     ("alloc/collections/btree_set/struct.BTreeSet.html", &["#insert-and-complex-keys"]),
 ];
 
+#[rustfmt::skip]
+const INTRA_DOC_LINK_EXCEPTIONS: &[(&str, &[&str])] = &[
```
Comment (GuillaumeGomez) [inline src/tools/linkchecker/main.rs:57]: Are you sure you want to run this check on html?

Comment (jyn514) [reply src/tools/linkchecker/main.rs:57]: Yes, there's no way for the linkchecker to tell if a link will be resolved by rustdoc or not. It has to look at the HTML to see whether the markdown was transformed into a link or left as-is.

File: compiler/rustc_error_codes/src/error_codes/E0660.md (hunk 1 above)
Comment (GuillaumeGomez) [reply compiler/rustc_error_codes/src/error_codes/E0660.md:12]: Oh you're right. Nevermind then!

File: src/tools/linkchecker/main.rs (hunk 4 above)
Comment (GuillaumeGomez) [reply src/tools/linkchecker/main.rs:57]: That sounds really risky. :-/

**Comment by jyn514** [comment]
Ok, this should be ready to go now. I updated two submodules that had broken links (the commit message has a `/compare` url).

**Comment by GuillaumeGomez** [comment]
This looks good to me. Let's wait for @Mark-Simulacrum's review too.

**Comment by Mark-Simulacrum** [comment]
I guess this seems fine, r=me.

It does feel like the sort of thing that would be better to fix in rustdoc, but I'm sure you're aware of that :)

**Comment by Mark-Simulacrum** [comment]
Oh, one note I wanted to make -- it look like this is now fixing links, could you update PR description?

**Comment by jyn514** [comment]
Updated!

> It does feel like the sort of thing that would be better to fix in rustdoc, but I'm sure you're aware of that :)

Yeah ... I'd like to have https://github.com/rust-lang/rust/issues/77200 and https://github.com/rust-lang/rust/issues/77199 but the rest of the rustdoc team disagrees with me (https://github.com/rust-lang/rust/pull/77276#issuecomment-700909230, various conversations on discord). In the meantime this should catch most of the broken links.

This is still a good change even if those issues were fixed though IMO - it protects against bugs in rustdoc (and hopefully makes them easier to find).
